src/Codec/ArrayCodec.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Codec;
+
+use MongoDB\Exception\UnsupportedValueException;
+
+use function array_map;
+use function is_array;
+
+/**
+ * Codec to recursively encode/decode values in arrays.
+ *
+ * @template-implements Codec<array, array>
+ */
+final class ArrayCodec implements Codec, KnowsCodecLibrary
+{
+    private ?CodecLibrary $library = null;
+
+    public function attachCodecLibrary(CodecLibrary $library): void
+    {
+        $this->library = $library;
+    }
+
+    /**
+     * @param mixed $value
+     * @psalm-assert-if-true array $value
+     */
+    public function canDecode($value): bool
+    {
+        return is_array($value);
+    }
+
+    /**
+     * @param mixed $value
+     * @psalm-assert-if-true array $value
+     */
+    public function canEncode($value): bool
+    {
+        return is_array($value);
+    }
+
+    /** @param mixed $value */
+    public function decode($value): array
+    {
+        if (! $this->canDecode($value)) {
+            throw UnsupportedValueException::invalidDecodableValue($value);
+        }
+
+        return array_map(
+            /**
+             * @param mixed $item
+             * @return mixed
+             */
+            fn ($item) => $this->getLibrary()->decodeIfSupported($item),
+            $value,
+        );
+    }
+
+    /**
+     * @param mixed $value
+     * @return mixed
+     * @psalm-return ($value is array ? array : $value)
+     */
+    public function decodeIfSupported($value)
+    {
+        return $this->canDecode($value) ? $this->decode($value) : $value;
+    }
+
+    /** @param mixed $value */
+    public function encode($value): array
+    {
+        if (! $this->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
+        }
+
+        return array_map(
+            /**
+             * @param mixed $item
+             * @return mixed
+             */
+            fn ($item) => $this->getLibrary()->encodeIfSupported($item),
```

Comment:
In essence, yes. In this case, `encode` relies on the value being an array, but doesn't know anything about the values it contains. Since the goal is just to traverse the array for encoding, we use `encodeIfSupported` on child elements.

The reason I split the API into `encode` throwing an exception and `encodeIfSupported` silently returning the original value if it isn't supported was to make this a conscious choice for the developer. In most cases we'll be using the `ifSupported` variants, but the `CodecCursor` introduced in #1140 uses `decode` as we always expect to receive a `Document` instance.